src/clients/api/mutations/useSupply.ts

```diff
@@ -0,0 +1,65 @@
+import { useMutation, MutationObserverOptions } from 'react-query';
+
+import { VTokenId } from 'types';
+import queryClient from 'clients/api/queryClient';
+import supplyBnb, { ISupplyBnbInput, SupplyBnbOutput } from 'clients/api/mutations/supplyBnb';
+import supply, { ISupplyInput, SupplyOutput } from 'clients/api/mutations/supply';
+import FunctionKey from 'constants/functionKey';
+import { useVTokenContract } from 'clients/contracts/hooks';
+import { getVTokenContract } from 'clients/contracts';
+
+interface IUseSupplyArgs {
+  assetId: VTokenId;
+  account: string | undefined;
+}
+
+export type UseSupplyParams =
+  | Omit<ISupplyInput, 'tokenContract' | 'assetId' | 'account'>
+  | Omit<ISupplyBnbInput, 'tokenContract' | 'assetId' | 'account'>;
+
+const getWeb3 = (params: UseSupplyParams) => {
+  if ('web3' in params) {
+    return params.web3;
+  }
+  return undefined;
+};
+
+const useSupply = (
+  { assetId, account }: IUseSupplyArgs,
+  // TODO: use custom error type https://app.clickup.com/t/2rvwhnt
+  options?: MutationObserverOptions<SupplyOutput | SupplyBnbOutput, Error, UseSupplyParams>,
+) => {
+  const vContract = useVTokenContract(assetId as VTokenId);
+  return useMutation(
+    [FunctionKey.SUPPLY, assetId],
+    params => {
+      const web3 = getWeb3(params);
+      const { amount } = params;
+      if (assetId === 'bnb' && web3) {
+        const vBnbContract = getVTokenContract('bnb', web3);
+        supplyBnb({
+          tokenContract: vBnbContract,
+          web3,
+          account,
+          amount,
+        });
+      }
+      return supply({
+        tokenContract: vContract,
+        account,
+        amount,
+      });
+    },
+    {
+      ...options,
+      onSuccess: (data, variables, context) => {
+        queryClient.invalidateQueries(FunctionKey.GET_VTOKEN_BALANCES_ALL);
+        if (options?.onSuccess) {
+          options.onSuccess(data, variables, context);
+        }
+      },
```

Comment:
A small thing: the params of `onSuccess` could change in a future update of React Query. For that reason, it's safer to just pass all the params:
```suggestion
      onSuccess: (...onSuccessParams) => {
        if (options?.onSuccess) {
          options.onSuccess(...onSuccessParams);
        }
        
        queryClient.invalidateQueries(FunctionKey.GET_VTOKEN_BALANCES_ALL);
      },
```